### ISSUE TITLE
[BUG] A running background task will block new requests from the OpenAPI view when using a metrics exporter middleware

### ISSUE THREAD

**Issue by DoronCohen** [tiangolo/fastapi#1355]
### To Reproduce

Steps to reproduce the behavior with a minimum self-contained file.

Replace each part with your own scenario:

1. Create a file with:

```Python
from time import sleep

from fastapi import FastAPI, BackgroundTasks
from starlette_exporter import PrometheusMiddleware, handle_metrics


app = FastAPI()
app.add_middleware(PrometheusMiddleware)
app.add_route("/metrics", handle_metrics)


def task(seconds):
    print("Starting task, sleep", seconds)
    sleep(seconds)
    print("Done task, sleep", seconds)


@app.get("/{seconds}")
async def test(seconds: int, background_tasks: BackgroundTasks):
    print(f"In route with seconds={seconds}")
    background_tasks.add_task(task, seconds=seconds)
    return f"{seconds} hello world!"
```

3. Open the OpenAPI docs view and click "Try It Now".
4. Send a request with 10 seconds (or something else but make it long enough).
5. The response will return immediately.
5. Immediately send another request and the request will block until the backgroud task will finish.

### Expected behavior

Requests shouldn't be blocked by background tasks

### Environment

- OS: Linux
- FastAPI Version 0.52.0
- Python version, 3.7.7

### Additional context

At first, I thought about opening this issue in starlette-prometheus package but then I tried it with starlette-exporter as you can see in the example and it still happens. I don't know why is it reproducing with the OpenAPI view but not via `curl` and I don't have the knowledge to investigate this issue.

**Comment by thomasmatecki** [comment]
Just for clarity, the request blocks at this line:
```
   background_tasks.add_task(task, seconds=seconds)
```

**Comment by DoronCohen** [comment]
No, the first request is handled correctly and fast. The second one is stuck in the OpenAPI interface and it doesnt even get to the route function. I tried both sync and async tasks.

**Comment by chris-allnutt** [comment]
@DoronCohen what happens if you register an async task instead of a sync one?  FastAPI says it handles both properly, but I'm wondering if that's incorrect, starlette only shows async examples.

**Comment by DoronCohen** [comment]
I tried both but they both behave the same way. I tried to think what does it have to do with the OpenAPI docs view. Still a mystery to me

**Comment by sebr74** [comment]
I have hit a similar issue, but I think this bug is actually a specific case of https://github.com/encode/starlette/issues/919. So a Starlette bug, not a FastAPI bug.

**Comment by emirmuminoglu** [comment]
I don't use any middleware but it happens on my app too. Why does it happen?

**Comment by jacopofar** [comment]
Same problem here, with the GZipMiddleware enabled the request is blocked until the background task finishes. If I disable it it works as expected.

The curious thing is that the log shows that the request was correctly processed, so I initially assumed it was a reverse proxy problem.

**Comment by amitripshtos** [comment]
Same here, could verify the issue with this code:
`from typing import Optional
import time
from fastapi import FastAPI, BackgroundTasks, Request
from fastapi.middleware.gzip import GZipMiddleware
app = FastAPI()

def lol():
    import time
    time.sleep(15)

app.add_middleware(GZipMiddleware, minimum_size=1000)

@app.middleware("http")
async def add_process_time_header(request: Request, call_next):
    start_time = time.time()
    response = await call_next(request)
    process_time = time.time() - start_time
    response.headers["X-Process-Time"] = str(process_time)
    return response

@app.get("/")
def read_root(background_tasks: BackgroundTasks):
    background_tasks.add_task(lol)
    return {"Hello": "World"}


@app.get("/items/{item_id}")
def read_item(item_id: int, q: Optional[str] = None):
    return {"item_id": item_id, "q": q}`

**Comment by rafalstapinski** [comment]
I'm experiencing a similar issue with the elastic `apm-agent-python` - running a background task blocks the request that initiated it from returning. When not adding the apm middleware, the response returns immediately as expected, and the background task runs separately until it finished.

**Comment by dmig-alarstudios** [comment]
@rafalstapinski https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor

**Comment by jiamo** [comment]
Same here.  Do a long task in background and start thread pool.
I am wondering the thread run in the long task is handing request too.

**Comment by chandrashekar-cohesity** [comment]
Is there a workaround for this? I am hitting the same issue.

**Comment by dbatten5** [comment]
has anyone had any luck with this? i'm experiencing the same thing, feels like quite an essential part of background tasks that it shouldn't block requests

**Comment by ycd** [comment]
It is related to https://github.com/encode/starlette/issues/919, for workaround you can use a custom handler.

**Comment by dbatten5** [comment]
@ycd that seems to relate to apps that use middleware extending `BaseHTTPMiddleware` and to my knowledge my app isn't using that:
```
(Pdb) app.middleware_stack
<starlette.middleware.errors.ServerErrorMiddleware object at 0x1095e3970> # ServerErrorMiddleware doesn't extend BaseHTTPMiddleware
(Pdb) app.user_middleware
[]
```
seems like my issue was actually due to my database. tried using postgres instead of sqlite and it worked fine

**Comment by juliusdanek** [comment]
+1 I am also seeing this error. Background tasks are blocking in the OpenAPI view when using Middlewares. A little annoying since I sometimes use the OpenAPI View for testing...

**Comment by mikheillomidze** [comment]
Same here, any updates?
